src/pip/_internal/vcs/__init__.py

```diff
@@ -227,6 +227,10 @@ def get_url_rev(self):
         assert '+' in self.url, error_message % self.url
         url = self.url.split('+', 1)[1]
         scheme, netloc, path, query, frag = urllib_parse.urlsplit(url)
+        if '#' in frag:
```

Comment:
It seems like this code may be more elegant if you pass `allow_fragments=False` to the call to `urlsplit()`, in part because you're effectively parsing the fragment yourself anyways.